### PR TITLE
Adman adapter

### PR DESCRIPTION
**Pull request by SmartyAdman**
New prebid-server adapter for connecting with adman platform

### CODE REVIEW

File: adapters/adman/params_test.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	}
+}
+
+var validParams = []string{
+	`{"TagID": "16"}`,
+}
+
+var invalidParams = []string{
+	`{"id": "123"}`,
+	`{"tagid": "123"}`,
```
Comment (SyntaxNode) [inline adapters/adman/params_test.go:44]: Please add a case for the proper name but an invalid JSON type, like:
```
`{"TagID": 16}`,
```

File: adapters/adman/usersync_test.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"testing"
+	"text/template"
+
+	"github.com/prebid/prebid-server/privacy"
+	"github.com/prebid/prebid-server/privacy/gdpr"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAdmanSyncer(t *testing.T) {
+	syncURL := "https://sync.admanmedia.com/pbs.gif?redir=http%3A%2F%2Flocalhost%3A8000%2Fsetuid%3Fbidder%3Dadman%26uid%3D%5BUID%5D"
```
Comment (SyntaxNode) [inline adapters/adman/usersync_test.go:13]: Please include the GDPR and GDPRConsent macros in the test. It's usually a literal copy and paste from config.go with the host hardcoded (the latter you have here already).

File: config/config.go [hunk 3]
```diff
@@ -502,6 +502,7 @@ func (cfg *Configuration) setDerivedDefaults() {
 	setDefaultUsersync(cfg.Adapters, openrtb_ext.BidderAdpone, "https://usersync.adpone.com/csync?redir="+url.QueryEscape(externalURL)+"%2Fsetuid%3Fbidder%3Dadpone%26gdpr%3D{{.GDPR}}%26gdpr_consent%3D{{.GDPRConsent}}%26uid%3D%7Buid%7D")
 	setDefaultUsersync(cfg.Adapters, openrtb_ext.BidderAdtelligent, "https://sync.adtelligent.com/csync?t=p&ep=0&redir="+url.QueryEscape(externalURL)+"%2Fsetuid%3Fbidder%3Dadtelligent%26gdpr%3D{{.GDPR}}%26gdpr_consent%3D{{.GDPRConsent}}%26uid%3D%7Buid%7D")
 	setDefaultUsersync(cfg.Adapters, openrtb_ext.BidderAdmixer, "https://inv-nets.admixer.net/adxcm.aspx?gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&redir=1&rurl="+url.QueryEscape(externalURL)+"%2Fsetuid%3Fbidder%3Dadmixer%26gdpr%3D{{.GDPR}}%26gdpr_consent%3D{{.GDPRConsent}}%26uid%3D%24%24visitor_cookie%24%24")
+	setDefaultUsersync(cfg.Adapters, openrtb_ext.BidderAdman, "https://sync.admanmedia.com/pbs.gif?redir="+url.QueryEscape(externalURL)+"%2Fsetuid%3Fbidder%3Dadman%26gdpr%3D{{.GDPR}}%26gdpr_consent%3D{{.GDPRConsent}}%26uid%3D%5BUID%5D")
```
Comment (SyntaxNode) [inline config/config.go:505]: Being in the EU, don't you need the GDPR and GDPRConsent macros resolved for your cookie sync endpoint? Right now, they're only resolved for the post back / redirect to Prebid Server.

File: config/config.go [hunk 4]
```diff
@@ -763,6 +764,7 @@ func SetupViper(v *viper.Viper, filename string) {
 	v.SetDefault("adapters.yieldmo.endpoint", "https://ads.yieldmo.com/exchange/prebid-server")
 	v.SetDefault("adapters.yieldone.endpoint", "https://y.one.impact-ad.jp/hbs_imp")
 	v.SetDefault("adapters.zeroclickfraud.endpoint", "http://{{.Host}}/openrtb2?sid={{.SourceId}}")
+	v.SetDefault("adapters.adman.endpoint", "http://eu-ams-1.admanmedia.com/?c=o&m=ortb")
```
Comment (SyntaxNode) [inline config/config.go:767]: Please add the default value in alphabetically.

File: exchange/adapter_map.go [hunk 5]
```diff
@@ -160,6 +161,7 @@ func newAdapterMap(client *http.Client, cfg *config.Configuration, infos adapter
 		openrtb_ext.BidderYieldmo:          yieldmo.NewYieldmoBidder(cfg.Adapters[string(openrtb_ext.BidderYieldmo)].Endpoint),
 		openrtb_ext.BidderYieldone:         yieldone.NewYieldoneBidder(cfg.Adapters[string(openrtb_ext.BidderYieldone)].Endpoint),
 		openrtb_ext.BidderZeroClickFraud:   zeroclickfraud.NewZeroClickFraudBidder(cfg.Adapters[string(openrtb_ext.BidderZeroClickFraud)].Endpoint),
+		openrtb_ext.BidderAdman:            adman.NewAdmanBidder(client, cfg.Adapters[string(openrtb_ext.BidderAdman)].Endpoint),
```
Comment (SyntaxNode) [inline exchange/adapter_map.go:164]: Please add the adapter mapping alphabetically.

File: openrtb_ext/bidders.go [hunk 6]
```diff
@@ -94,6 +94,7 @@ const (
 	BidderYieldmo          BidderName = "yieldmo"
 	BidderYieldone         BidderName = "yieldone"
 	BidderZeroClickFraud   BidderName = "zeroclickfraud"
+	BidderAdman			   BidderName = "adman"
```
Comment (SyntaxNode) [inline openrtb_ext/bidders.go:97]: Ah. Looks like you need to run `go fmt` on the PR.

File: usersync/usersyncers/syncer.go [hunk 7]
```diff
@@ -75,6 +76,7 @@ import (
 func NewSyncerMap(cfg *config.Configuration) map[openrtb_ext.BidderName]usersync.Usersyncer {
 	syncers := make(map[openrtb_ext.BidderName]usersync.Usersyncer, len(cfg.Adapters))
 
+	insertIntoMap(cfg, syncers, openrtb_ext.BidderAdman, adman.NewAdmanSyncer)
```
Comment (SyntaxNode) [inline usersync/usersyncers/syncer.go:79]: Alphabetically please.

File: usersync/usersyncers/syncer_test.go [hunk 8]
```diff
@@ -73,6 +73,7 @@ func TestNewSyncerMap(t *testing.T) {
 			string(openrtb_ext.BidderYieldmo):          syncConfig,
 			string(openrtb_ext.BidderYieldone):         syncConfig,
 			string(openrtb_ext.BidderZeroClickFraud):   syncConfig,
+			string(openrtb_ext.BidderAdman):            syncConfig,
 		},
```
Comment (SyntaxNode) [inline usersync/usersyncers/syncer_test.go:77]: Alphabetically please.

**Comment by SmartyAdman** [comment]
Thank you for deep code review, i've made all changes according to your comments

**Comment by SmartyAdman** [comment]
> Travis, that not only runs test cases but also tests for race conditions and the use of `gofmt`, is complaining precisely about the latter:
> 
> ```
> ./validate.sh --nofmt --cov --race 10
> 
> gofmt needs to be run, 1 files have issues.  Below is a list of files to review:
> 
> openrtb_ext/bidders.go
> 
> The command "./validate.sh --nofmt --cov --race 10" exited with 1.
> ```
> 
> Can you please correct?

@guscarreon I've fixed this.

File: adapters/adman/adman.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+// MakeRequests create bid request for adman demand
+func (a *AdmanAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var admanExt openrtb_ext.ExtImpAdman
+	var err error
+
+	var adapterRequests []*adapters.RequestData
+
+	for _, imp := range request.Imp {
+		reqCopy := *request
```
Comment (guscarreon) [inline adapters/adman/adman.go:39]: Given that `reqCopy.Imp` is the only `reqCopy` field that gets modified at every loop, I believe it is safe and a bit more optimal to make this copy outside the `for` loop.
```
 31   func (a *AdmanAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
 32       var errs []error
 33       var admanExt openrtb_ext.ExtImpAdman
 34       var err error
 35  
 36       var adapterRequests []*adapters.RequestData
 37  
    +     reqCopy := *request
 38       for _, imp := range request.Imp {
 39 -         reqCopy := *request
 40           reqCopy.Imp = []openrtb.Imp{imp}
 41  
 42   *-- 17 lines: var bidderExt adapters.ExtImpBidder--------------------------------------------------------------------------
 59           errs = append(errs, errors...)
 60       }
 61       return adapterRequests, errs
 62   }
adapters/adman/adman.go
```

**Comment by guscarreon** [comment]
@SmartyAdman thank you for addressing the small optimization issue. Can we please add test cases to reach at least a 90% test coverage?

**Comment by SmartyAdman** [comment]
@guscarreon Sorry for delays, finally i've added some test and coverage now around 92%

**Pull request merged**